src/main/java/org/opensrp/repository/postgres/ClientFormRepositoryImpl.java

```diff
@@ -0,0 +1,154 @@
+package org.opensrp.repository.postgres;
+
+import org.springframework.lang.NonNull;
+import org.springframework.lang.Nullable;
+import org.apache.http.util.TextUtils;
+import org.opensrp.domain.IdVersionTuple;
+import org.opensrp.domain.postgres.ClientForm;
+import org.opensrp.domain.postgres.ClientFormMetadata;
+import org.opensrp.repository.ClientFormRepository;
+import org.opensrp.repository.postgres.mapper.custom.CustomClientFormMapper;
+import org.opensrp.repository.postgres.mapper.custom.CustomClientFormMetadataMapper;
+import org.opensrp.service.ClientFormService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.transaction.InvalidTransactionException;
+import java.util.List;
+
+@Repository
+public class ClientFormRepositoryImpl extends BaseRepositoryImpl<ClientForm> implements ClientFormRepository {
+
+	@Autowired
+	private CustomClientFormMapper clientFormMapper;
+
+	@Autowired
+	private CustomClientFormMetadataMapper clientFormMetadataMapper;
+
+	@Override
+	public ClientForm get(long id) {
+		return clientFormMapper.selectClientFormById(id);
+	}
+
+	@Override
+	public int countClientFormByFormIdentifier(@NonNull String formIdentifier) {
+		return clientFormMetadataMapper.countClientFormMetadataByFormIdentifier(formIdentifier);
+	}
+
+	@Override
+	public ClientFormMetadata getClientFormMetadata(@NonNull String formVersion, @NonNull String formIdentifier) {
+		return clientFormMetadataMapper.selectClientFormMetadataByFormVersionAndIdentifier(formVersion, formIdentifier);
+	}
+
+	@Override
+	public List<IdVersionTuple> getAvailableClientFormVersions(@NonNull String formIdentifier) {
+		return clientFormMetadataMapper.getAvailableClientFormVersions(formIdentifier);
+	}
+
+	@Override
+	public ClientFormMetadata getFormMetadata(long id) {
+		return clientFormMetadataMapper.selectByPrimaryKey(id);
+	}
+
+	@Nullable
+	@Transactional(rollbackFor = { InvalidTransactionException.class })
+	@Override
+	public ClientFormService.CompleteClientForm create(@NonNull ClientForm clientForm,
+			@NonNull ClientFormMetadata clientFormMetadata) throws InvalidTransactionException {
+		Long clientFormId = clientForm.getId();
+		Long clientFormMetadataId = clientFormMetadata.getId();
+		if ((clientFormId != null && clientFormId != 0) || (clientFormMetadataId != null && clientFormMetadataId != 0)) {
+			logger.error("ClientForm & ClientFormMetadata Id must be NULL or 0", new Exception());
+			return null;
```

Comment:
Should maybe throw an `IllegalArgumentException`

Reply:
Cool